tree/30_generic_methods/sharedfile_to_node.cf

```diff
@@ -0,0 +1,97 @@
+#####################################################################################
+# Copyright 2017 Normation SAS
+#####################################################################################
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, Version 3.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#####################################################################################
+
+# @name Sharedfile to node
+# @description This method shares a file with another Rudder node
+# @documentation This method share a file with another rudder node using a unique file identifier.
+#  The file will be kept on the policy server and transmitted to the destination node's policy server if it is different.
+#  It will be kept on this server for the destination node to download as long as it is not replaced by a new
+#  file with the same id or remove by expiration of the TTL. 
+#  The TTL can be a number of second or a string of the long form "1day 2hours 3minute 4seconds" or abbreviated in the form "5h 3s"
+#  Example: 
+# ```
+# sharedfile_to_node("73570beb-2d4a-43d2-8ffc-f84a6817849c", "application.properties", "/srv/db/application.properties", "356days")
+# ```
+#
+# @parameter target_uuid  which node to share the file with
+# @parameter file_id      unique name that will be used to identify the file on the receiver
+# @parameter file_path    path of the file to share
+# @parameter ttl          time to keep the file on the policy server in seconds or in human readable form (see long description)
+# 
+# @class_prefix sharedfile_to_node
+# @class_parameter file_id
+
+bundle agent sharedfile_to_node(target_uuid, file_id, file_path, ttl)
+{
+  vars:
+      "old_class_prefix"  string => canonify("sharedfile_to_node_${file_id}");
+      "promisers"          slist => { @{this.callers_promisers}, cf_null }, policy => "ifdefined";
+      "class_prefix"      string => canonify(join("_", "promisers"));
+      "args"              slist  => { "${target_uuid}", "${file_id}", "${file_path}", "${ttl}" };
+
+      "canonified_path"   string => canonify("${file_path}");
+      "url"               string => "https://${def.policy_server}/rudder/relay-api/shared-files/${target_uuid}/${g.uuid}/${file_id}";
+      # TODO store this hash in the signature
+      "hash"              string => execresult("sha512sum ${file_path} | cut -d ' ' -f 1", "useshell");
```

Comment:
We may want to fail early with a clear error message when we can't find rudder-sign (vanilla CFEngine on Linux, Windows agents, etc.)